examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,373 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Entrypoint that runs the Composer trainer on a provided YAML hparams file.
+Specifically designed for the NLP use case, allowing pre-training and fine-tuning on 
+downstream tasks to be handled within one script.
+
+Example that pretrains a BERT:
+    >>> composer examples/run_nlp_trainer.py 
+    -f composer/yamls/models/bert-base.yaml 
+    --training_scheme pretrain 
+
+Example that pretrains and finetunes a BERT:
+    >>> composer examples/run_nlp_trainer.py 
+    -f composer/yamls/models/bert-base.yaml 
+    --training_scheme all 
+    --save_folder ~/checkpoints
+
+Example that finetunes a pretrained BERT:
+
+    >>> composer examples/run_nlp_trainer.py
+    -f composer/yamls/models/bert_base.yaml
+    --training_scheme finetune
+    --load_path ~/checkpoints
+"""
+import argparse
+import dataclasses
+import logging
+import multiprocessing as mp
+import os
+import subprocess
+import sys
+import warnings
+from concurrent.futures import Future
+from concurrent.futures import ProcessPoolExecutor as Pool
+from functools import partial
+from multiprocessing.pool import ThreadPool
+from typing import List, Optional, Union
+
+import boto3
+import torch
+import yahp as hp
+from tabulate import tabulate
+
+import composer
+from composer.loggers.object_store_logger import ObjectStoreLogger
+from composer.loggers.wandb_logger import WandBLogger
+from composer.metrics.glue import GlueState
+from composer.trainer.devices.device_gpu import DeviceGPU
+from composer.trainer.trainer_hparams import TrainerHparams
+from composer.utils.entrypoint import _warning_on_one_line
+from composer.utils.object_store.object_store_hparams import S3ObjectStoreHparams
+from composer.utils.object_store.s3_object_store import S3ObjectStore
+
+# TODO: reset all configs and training lengths back to default
+
+''' Generate a multiprocessing queue to store queue_size GPU IDs. '''
+def init_cuda_queue(queue_size: int) -> mp.Queue:
```

Comment:
Based on our offline conversation, do we need the queue? It seems a bit heavy to have to use a queue to pass a unqiue id to each worker. Instead, could each worker call `multiprocessing.current_process()` (https://docs.python.org/3/library/multiprocessing.html#multiprocessing.current_process) within `init_cuda_env`?

Reply:
From my understanding, there is no way to extract a unique ID from mp.current_process(), which returns a Process object. The only way is to either somehow map process IDs to their GPU equivalents, or parse the worker number out of the Process name, neither of which seem to be less heavy or cleaner than the queue.

Reply:
Discussed offline -- will keep the queue and add justification in documentation.